src/borsh.rs

```diff
@@ -0,0 +1,167 @@
+#![cfg_attr(docsrs, doc(cfg(feature = "borsh")))]
+
+use alloc::vec::Vec;
+use core::hash::BuildHasher;
+use core::hash::Hash;
+use core::mem::size_of;
+
+use borsh::error::ERROR_ZST_FORBIDDEN;
+use borsh::io::{Error, ErrorKind, Read, Result, Write};
+use borsh::{BorshDeserialize, BorshSerialize};
+#[cfg(feature = "borsh-schema")]
+use ::{
+    alloc::collections::btree_map::BTreeMap,
+    alloc::format,
+    borsh::schema::{add_definition, Declaration, Definition},
+    borsh::BorshSchema,
+};
+
+use crate::map::IndexMap;
+use crate::set::IndexSet;
+
+impl<K, V, H> BorshSerialize for IndexMap<K, V, H>
+where
+    K: BorshSerialize + Ord,
+    V: BorshSerialize,
+    H: BuildHasher,
+{
+    #[inline]
+    fn serialize<W: Write>(&self, writer: &mut W) -> Result<()> {
+        check_zst::<K>()?;
+
+        let mut vec = self.iter().collect::<Vec<_>>();
+        vec.sort_by(|(a, _), (b, _)| a.cmp(b));
```

Comment:
it's part of the borsh spec. encoded values must be sorted to guarantee that they have the same representation, and are therefore safe for hashing in replicated state machines

Reply:
Do you have a reference?

It is also possible to think of `IndexMap` as a `Vec` with map-like methods -- which is actually close to how it is implemented -- and I can see at a glance that `BorshSerialize` doesn't change order for `Vec`. Maybe the problem is that `PartialEq`/`Eq` for `IndexMap` doesn't respect order? That's why we don't implement `Hash` either...

At the very least, we need to document that sorting effect. Note for example #156 with serde, where *some* backends don't preserve map order, and we added the `serde_seq` module to opt into strict sequence order. We also use that sequence implementation for serializing `map::Slice`. (`Slice` doesn't support deserializing.) These kind of alternatives might be useful to offer with borsh too for when users do care about the original order.

Reply:
borsh's specs are documented here => https://borsh.io/ (check the specification tab, under hashmaps and hashsets)

I see what you mean. hmm I honestly hadn't thought about this; it is true that by serializing values like this, we'll lose the original order of the `Index{Map,Set}` data structures. `IndexMap` is a leaky map abstraction (in the sense that order matters), which kind of throws off the assumptions about hashmaps in borsh.

I'll take a look at `serde_seq` tomorrow. my first instinct (now), though, is to create two adapters to serialize the map with, and pick one as the default (obv documenting this behavior). I'm guessing this is how `serde` is handled, too.

unfortunately, to preserve the order of indexmaps with borsh, we need to serialize the table of indexes along with the buffer of key-value pairs, but it might be better to make this the default behavior. the alternative adapter could have a lean (i.e. without the index table) serialization method that reflected the current implementation.

Reply:
That spec seems rather loose, not getting into stuff like what you said about "encoded map and set values must be sorted to guarantee that they have the same representation, and are therefore safe for hashing in replicated state machines." The only bit that I see is:
> all unordered containers (hashmap/hashset) are ordered in lexicographic order by key (in tie breaker case on value); 
but `IndexMap` **is** ordered, though with some oddity like what I said about `Eq`.

> unfortunately, to preserve the order of indexmaps with borsh, we need to serialize the table of indexes along with the buffer of key-value pairs

Why? Couldn't it be serialized as if it were just a `Vec<(K, V)>`? From what you've implemented, I don't see any way that borsh could "know" or care that `IndexMap` has map-like behavior at all.

Reply:
disregard this entire conversation. haha
I've pushed your original suggestion